src/Juvix/Compiler/Backend/Geb/Analysis/TypeChecking.hs

```diff
@@ -0,0 +1,283 @@
+              _inferenceEnvTypeInfo = Just aType
+            }
+        )
+    )
+    (check (l ^. lambdaBody) bType)
+  return $
+    ObjectHom $
+      Hom
+        { _homDomain = aType,
+          _homCodomain = bType
+        }
+
+inferObjectPair :: InferEffects r => Pair -> Sem r Object
+inferObjectPair pair = do
+  let lType = pair ^. pairLeftType
+      rType = pair ^. pairRightType
+  check (pair ^. pairLeft) lType
+  check (pair ^. pairRight) rType
+  return $
+    ObjectProduct
+      Product
+        { _productLeft = lType,
+          _productRight = rType
+        }
+
+inferObjectCase :: InferEffects r => Case -> Sem r Object
+inferObjectCase c = do
+  let aType = c ^. caseLeftType
+      bType = c ^. caseRightType
+      vType =
+        ObjectCoproduct $
+          Coproduct
+            { _coproductLeft = aType,
+              _coproductRight = bType
+            }
+      cType = c ^. caseCodomainType
+      leftType =
+        ObjectHom $
+          Hom
+            { _homDomain = aType,
+              _homCodomain = cType
+            }
+      rightType =
+        ObjectHom $
+          Hom
+            { _homDomain = bType,
+              _homCodomain = cType
+            }
+  check (c ^. caseOn) vType
+  check (c ^. caseLeft) leftType
+  check (c ^. caseRight) rightType
+  return cType
+
+inferObjectFirst :: InferEffects r => First -> Sem r Object
+inferObjectFirst p = do
+  let leftType = p ^. firstLeftType
+      rightType = p ^. firstRightType
+      pairType =
+        ObjectProduct $
+          Product
+            { _productLeft = leftType,
+              _productRight = rightType
+            }
+  check (p ^. firstValue) pairType
+  return leftType
+
+inferObjectSecond :: InferEffects r => Second -> Sem r Object
+inferObjectSecond p = do
+  let leftType = p ^. secondLeftType
+      rightType = p ^. secondRightType
+      pairType =
+        ObjectProduct $
+          Product
+            { _productLeft = leftType,
+              _productRight = rightType
+            }
+  check (p ^. secondValue) pairType
+  return rightType
+
+inferObjectVar :: InferEffects r => Var -> Sem r Object
+inferObjectVar v = do
+  ctx <- asks (^. inferenceEnvContext)
+  return $ Context.lookup (v ^. varIndex) ctx
+
+inferObjectBinop :: InferEffects r => Binop -> Sem r Object
+inferObjectBinop opApp = do
+  let outTy = objectBinop opApp
+      leftArg = opApp ^. binopLeft
+      rightArg = opApp ^. binopRight
+      args = [leftArg, rightArg]
+
+  case opApp ^. binopOpcode of
+    OpAdd -> do
+      checkListSameType args ObjectInteger
+      return outTy
+    OpSub -> do
+      checkListSameType args ObjectInteger
+      return outTy
+    OpMul -> do
+      checkListSameType args ObjectInteger
+      return outTy
+    OpDiv -> do
+      checkListSameType args ObjectInteger
+      return outTy
+    OpMod -> do
+      checkListSameType args ObjectInteger
+      return outTy
+    OpEq -> do
+      checkSameType args
```

Comment:
It is yet unspecified what the types of arguments to Eq can be (depends on GEB - Eq is not yet there), but probably only integers, maybe some other primitive types.